### ISSUE TITLE
Degraded block IO performace for sequential writes on KVM

### ISSUE THREAD

**Issue by russell-islam** [cloud-hypervisor/cloud-hypervisor#6786]
**Describe the bug**

**To Reproduce**
Steps to reproduce the behaviour:

**Version**
latest main

Did you build from source, if so build command line (e.g. features): cargo build --all --release

**VM configuration**

What command line did you run (or JSON config data):

./cloud-hypervisor/target/release/cloud-hypervisor \
	--kernel ./linux-cloud-hypervisor/arch/x86/boot/compressed/vmlinux.bin \
	--disk path=focal-server-cloudimg-amd64.raw,direct=on \
	--cpus boot=1 --memory size=2G \
	--cmdline "root=/dev/vda1 console=ttyS0" --log-file clh.log -vv\
    --net fd=3,mac=$mac 3<>$"$tapdevice"

Guest OS version details:
5.15.12+
Host OS version details:
6.8.0-45-generic
**Logs**

**Qemu:**
qemu-system-x86_64 -machine q35,accel=kvm,nvdimm=on -smp 1,cores=1,threads=1,sockets=1,maxcpus=1 -cpu host,pmu=off -m 2G -kernel ./linux-cloud-hypervisor/arch/x86/boot/compressed/vmlinux.bin -driv
e file=focal-server-cloudimg-amd64.raw,format=raw,if=virtio,cache=directsync -append "root=/dev/vda1 console=ttyS0" -nographic

**benchmark command:**
sysbench fileio run --threads=1 --file-test-mode=seqwr --file-num=4 --file-total-size=256m

CH:
File operations:
    reads/s:                      0.00
    writes/s:                     728.61
    fsyncs/s:                     29.34

Throughput:
    read, MiB/s:                  0.00
    written, MiB/s:               11.38

General statistics:
    total time:                          10.0123s
    total number of events:              7590

Latency (ms):
         min:                                    0.00
         avg:                                    1.32
         max:                                  234.04
         95th percentile:                        0.08
         sum:                                10001.76

Threads fairness:
    events (avg/stddev):           7590.0000/0.00
    execution time (avg/stddev):   10.0018/0.00

Qemu:

File operations:
    reads/s:                      0.00
    writes/s:                     1228.04
    fsyncs/s:                     49.52

Throughput:
    read, MiB/s:                  0.00
    written, MiB/s:               19.19

General statistics:
    total time:                          10.0914s
    total number of events:              12896

Latency (ms):
         min:                                    0.00
         avg:                                    0.78
         max:                                  135.75
         95th percentile:                        0.04
         sum:                                10078.53

Threads fairness:
    events (avg/stddev):           12896.0000/0.00
    execution time (avg/stddev):   10.0785/0.00

**Comment by russell-islam** [comment]
Analyzing with perf tool
![collage](https://github.com/user-attachments/assets/5f9bd8e3-5016-4852-b8bd-0b04f3d128c0)

**Comment by russell-islam** [comment]
Used the following patch(adding seg_max), does not make any difference
![seg_max](https://github.com/user-attachments/assets/a33537c3-a668-475a-833e-811af8316157)

**Comment by russell-islam** [comment]
Based on @liuw 's suggestion I was looking into the io_uring code to see if the code notifies io_uring everytime a request is parsed, I saw for every request we send the data to io_using and sync/submit. I am not sure if we can batch them. I tried to get some numbers from the process_queue_submit function to see how can we improve if we batch them. Following is the numbers:

```

# of requests   |     : Frequency
----------------|----------------
0                        778
----------------|----------------
1                        553
----------------|----------------
2                        64
----------------|----------------
8                        25
----------------|----------------
5                        1
----------------|----------------
3                        1
----------------|----------------
Current push: 889 can be: 644 if batched them not sure though
```

**Comment by russell-islam** [comment]
@rbradford @likebreath @liuw I would seek some advice.

**Comment by jinankjain** [comment]
Are you using io-ring backend for the QEMU side as well. I am asking because I don't see any iou worker in the perf result you attached?

**Comment by russell-islam** [comment]
> Are you using io-ring backend for the QEMU side as well. I am asking because I don't see any iou worker in the perf result you attached?

I have updated the image. Do you find any symbol at least?

**Comment by liuw** [comment]
> Are you using io-ring backend for the QEMU side as well. I am asking because I don't see any iou worker in the perf result you attached?

The QEMU command line has `aio=io_uring`.

**Comment by russell-islam** [comment]
@rbradford  Can we do execute_async with any array of requests?
I am not sure if we could do with current framework or how much we would benefit from?

**Comment by russell-islam** [comment]
> You tried adding SEG_MAX support. What prompted you to do that? Did you compare all the features and found that one to be missing? Are there other discrepancies you can find?

Yes. I did mix and match with the missing FEATURES in CLH but present in QEMU. Adding SEG_MAX did not make any differences.

**Comment by russell-islam** [comment]
I tried to batch the request to io_uring and improved very little in performance.

`Threads started!


File operations:
    reads/s:                      0.00
    writes/s:                     803.53
    fsyncs/s:                     32.34

Throughput:
    read, MiB/s:                  0.00
    written, MiB/s:               12.56

General statistics:
    total time:                          10.0753s
    total number of events:              8422

Latency (ms):
         min:                                    0.00
         avg:                                    1.20
         max:                                  593.43
         95th percentile:                        0.06
         sum:                                10065.02

Threads fairness:
    events (avg/stddev):           8422.0000/0.00
    execution time (avg/stddev):   10.0650/0.00

`

**Comment by rbradford** [comment]
@russell-islam There are some details on how to get nice perf reports including symbols in docs/profiling.md

tl;dr:
`cargo clean && RUSTFLAGS='-C force-frame-pointers=y' cargo build --profile profiling`
